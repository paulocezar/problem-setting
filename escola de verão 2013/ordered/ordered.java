import java.util.Random;
import java.util.Scanner;

public class ordered {
    
    static Random random = new Random();
    
    public static class Treap {
        
        int key;
        int first;
        int last;
        int sz;
        boolean inc;
        boolean dec;
        
        long pr;
        
        Treap left;
        Treap right;
        
        Treap(int value) {
            key = value;
            left = right = null;
            first = last = value;
            inc = dec = true;
            sz = 1;
            pr = random.nextLong();
        }
        
        void fix() {
            sz = 1 + getSize(left) + getSize(right);
            
            boolean linc = true, ldec = true;
            boolean rinc = true, rdec = true;
            int llast = key, rfirst = key;
            
            if (left != null) {
                first = left.first;
                llast = left.last;
                linc = left.inc;
                ldec = left.dec;
            } else {
                first = key;
            } 
            
            if (right != null) {
                last = right.last;
                rfirst = right.first;
                rinc = right.inc;
                rdec = right.dec;
            } else {
                last = key;
            }
            
            inc = linc && (llast <= key) && (key <= rfirst) && rinc;
            dec = ldec && (llast >= key) && (key >= rfirst) && rdec;
        }
    }
    
    static int getSize(Treap root) {
        return (root == null) ? 0 : root.sz;
    }    
    
    static class TreapPair {
        Treap left;
        Treap right;
        
        TreapPair(Treap left, Treap right) {
            this.left = left;
            this.right = right;
        }
    } 
    
    static TreapPair split(Treap root, int pos) {
        if (root == null) {
            return new TreapPair(null, null);
        }

        if (pos <= getSize(root.left)) {
            TreapPair sub = split(root.left, pos);
            root.left = sub.right;
            root.fix();
            sub.right = root;
            return sub;
        } else {
            TreapPair sub = split(root.right, pos - getSize(root.left) - 1);
            root.right = sub.left;
            root.fix();
            sub.left = root;
            return sub;
        }
    }
    
    static Treap merge(Treap left, Treap right) {
        if (left == null)
            return right;
        if (right == null)
            return left;
        if (left.pr > right.pr) {
            left.right = merge(left.right, right);
            left.fix();
            return left;
        } else {
            right.left = merge(left, right.left);
            right.fix();
            return right;
        }
    }
    
    static Treap insert(Treap root, int pos, int x) {
        TreapPair t = split(root, pos);
        return merge(merge(t.left, new Treap(x)), t.right);
    }
    
    static Treap remove(Treap root, int pos) {
        TreapPair t = split(root, pos);
        return merge(t.left, split(t.right, pos + 1 - getSize(t.left)).right);
    }
    
    static Treap modify(Treap root, int a, int newkey) {
      TreapPair t1 = split(root, a + 1);
      TreapPair t2 = split(t1.left, a);
      t2.right.key = newkey;
      t2.right.fix();
      return merge(merge(t2.left, t2.right), t1.right);
    }
    
    static int ith(Treap root, int k) {
        if (k < getSize(root.left))
            return ith(root.left, k);
        else if (k > getSize(root.left))
            return ith(root.right, k - getSize(root.left) - 1);
        return root.key;
    }
    
    static class TreapAndResult {
      Treap treap;
      String value;

      TreapAndResult(Treap t, String value) {
        this.treap = t;
        this.value = value;
      }
    }
    
    static TreapAndResult query(Treap root, int a, int b) {
      
      TreapPair t1 = split(root, b + 1);
      TreapPair t2 = split(t1.left, a);
      boolean inc = t2.right.inc;
      boolean dec = t2.right.dec;
      String res = "NONE";
      
      if (inc && dec) res = "ALL EQUAL";
      else if (inc) res = "NON DECREASING";
      else if (dec) res = "NON INCREASING";

      return new TreapAndResult(merge(merge(t2.left, t2.right), t1.right), res);
    }
    
    static void print(Treap root) {
      if (root == null)
        return;
      print(root.left);
      System.err.print(root.key + " ");
      print(root.right);
    }
    
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int N, Q, ox, oy, cmd, x, y;
        
        while (in.hasNextInt()) {
            N = in.nextInt();
            
            Treap treap = null;
            for (int i = 0; i < N; ++i) {
                x = in.nextInt();
                treap = insert(treap, i, x);
            }

            Q = in.nextInt();
            while (Q-->0) {
                cmd = in.nextInt();
                switch (cmd) {
                    case 0:
                        x = in.nextInt(); y = in.nextInt();
                        ox = ith(treap, x-1);
                        oy = ith(treap, y-1);
                        treap = modify(treap, x-1, oy);
                        treap = modify(treap, y-1, ox);
                    break;
                    case 1:
                        x = in.nextInt(); y = in.nextInt();
                        treap = modify(treap, x-1, y);
                    break;
                    case 2:
                        x = in.nextInt(); y = in.nextInt();
                        treap = insert(treap, x-1, y);
                    break;
                    case 3:
                        x = in.nextInt();
                        treap = remove(treap, x-1);
                    break;
                    case 4:
                        x = in.nextInt(); y = in.nextInt();
                        TreapAndResult tr = query(treap, x-1, y-1);
                        treap = tr.treap;
                        String v = tr.value;
                        System.out.println(v);
                    break;
                }
            }
        }
    }
}