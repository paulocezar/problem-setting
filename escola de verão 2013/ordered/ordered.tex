\documentclass[12pt]{article}

\usepackage[a4paper,top=2.0cm,bottom=2.0cm,left=2.0cm,right=2.0cm,nohead,nofoot]{geometry}
\usepackage[english,brazil]{babel}
\usepackage[T1]{fontenc}
\usepackage[ansinew]{inputenc}
\usepackage{indentfirst}
\usepackage{graphicx}
\usepackage{amsmath}

\begin{document}
\pagestyle{empty}

%----------------------------------------------------------------
\section*{Is it ordered?}
Chavaska likes to play with integer sequences. He has a sequence $A$ consisting of $N$ integers that he modifies and analyzes. Particularly he is interested in the ordering of some contiguous subsequences.

He explained to Kabralouco how he is having fun and invited him to play. Kabralouco wants to play but as he can't think as fast as Chavaska and doesn't like to stay behind, he decided to cheat and now asks your help creating a program capable of quickly dealing with the following operations:

\begin{itemize}
\item $0\ X\ Y$ -- Swap the elements at positions X and Y.
\item $1\ X\ Y$ -- Change the value of element at position X to Y.
\item $2\ X\ Y$ -- Insert element Y at position X.
\item $3\ X$ -- Remove element at position X.
\item $4\ X\ Y$ -- Report the ordering of the elements on (the subsequence) A[X...Y].
\begin{itemize}
	\item ``ALL EQUAL'' -- If $A[i] = A[i+1]$ for all $i$ in $[X, Y-1]$
	\item ``NON DECREASING'' -- If $A[i] \leq A[i+1]$ for all $i$ in $[X,Y-1]$ and $A[i] \neq A[i+1]$ for some $i$ in $[X,Y-1]$
	\item ``NON INCREASING'' -- If $A[i] \geq A[i+1]$ for all $i$ in $[X,Y-1]$ and $A[i] \neq A[i+1]$ for some $i$ in $[X,Y-1]$
	\item ``NONE'' -- If none of the above orderings happens.
\end{itemize}
\end{itemize}

\subsection*{\underline{Input}}
There might be several test cases. The first line of each test case begins with an integer $1 \leq N \leq 10^4$, the amount of numbers on the initial sequence.
On the next line there are $N$ integers $|A_i| \leq 10^9$ ($1 \leq i \leq N$).
The next line contains an integer $1 \leq Q \leq 10^5$, the number of operations that must be executed.
The following $Q$ lines are the operations, and will have space separated integers as described above. 

\subsection*{\underline{Output}}
Your program should output one line for each query (``4 X Y''), answering whether the subsequence A[X...Y] is NON INCREASING, NON DECREASING, ALL EQUAL or NONE as explained above.

\subsection*{\underline{Example}}
\begin{center}\tt
\begin{tabular}{|l|}
\hline
Input: \\
\hline
7\\
15 6 24 48 56 6 2\\
11\\
4 3 5\\
4 5 7\\
4 1 7\\
0 3 6\\
4 2 5\\
2 3 6\\
4 2 4\\
1 3 13\\
4 2 4\\
3 3\\
4 2 4\\
\hline
\end{tabular}\qquad\qquad
\begin{tabular}{|l|}
\hline
Output: \\
\hline
NON DECREASING\\
NON INCREASING\\
NONE\\
NON DECREASING\\
ALL EQUAL\\
NONE\\
NON DECREASING\\
\hline
\multicolumn{1}{c}{}\\[86pt]
\end{tabular}
\end{center}

\end{document}

