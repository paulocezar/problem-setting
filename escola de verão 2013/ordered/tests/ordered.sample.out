NON DECREASING
NON INCREASING
NONE
NON DECREASING
ALL EQUAL
NONE
NON DECREASING
